packages/typespec-azure-portal-core/src/decorators.ts

```diff
@@ -0,0 +1,253 @@
+      let filePath = resolvePath(dirPath, argQueryPathValue); //if given path is fullpath, it will return the fullPath
+      if (filePath && argQueryPath && argQueryPathValue) {
+        filePath = normalizePath(filePath);
+        if (!fs.existsSync(filePath)) {
+          reportDiagnostic(program, {
+            code: "fileNotFound",
+            messageId: "browseargQuery",
+            target,
+          });
+        }
+        (argQueryPath.type as StringLiteral).value = filePath;
+      }
+    }
+  }
+  program.stateMap(PortalCoreKeys.browse).set(target, options);
+}
+
+export function getBrowse(program: Program, target: Type) {
+  return program.stateMap(PortalCoreKeys.browse).get(target);
+}
+
+export function getBrowseArgQuery(program: Program, target: Type) {
+  const browse = getBrowse(program, target);
+  return browse.properties.get("argQuery");
+}
+
+export function isARMResource(
+  program: Program,
+  target: Model,
+  decoratorName: "browse" | "about" | "marketplaceOffer"
+) {
+  if (
+    getArmResourceKind(target) !== ("Tracked" as ArmResourceKind) &&
+    getArmResourceKind(target) !== ("Proxy" as ArmResourceKind)
+  ) {
+    reportDiagnostic(program, {
+      code: "invalidUsageDecorator",
+      messageId: decoratorName,
+      target,
+    });
+  }
+}
+/**
+ * This is a About decorator that will be used to define icon, keywords and learnMoreDocs.
+ * @param target The model that is being decorated.
+ * @param options AboutOptions of the property.
+ */
+export function $about(context: DecoratorContext, target: Model, options: AboutOptions) {
+  const { program } = context;
+  validateDecoratorUniqueOnNode(context, target, $about);
+  isARMResource(program, target, "about");
+  if (options && (options as Model).properties) {
+    const icon = (options as Model).properties.get("icon");
+    if (icon) {
+      if (icon.type.kind === "Model") {
+        const sourceLocation = getSourceLocation(target);
+        const dirPath =
+          sourceLocation && sourceLocation.file.path && getDirectoryPath(sourceLocation.file.path);
+        let iconPath = icon.type && (icon.type as Model).properties.get("filePath");
+        const iconPathValue = iconPath && iconPath.type && (iconPath.type as StringLiteral).value;
+        let filePath = resolvePath(dirPath, iconPathValue); //if given path is fullpath, it will return the fullPath
+        if (filePath && iconPath && iconPathValue) {
+          filePath = normalizePath(filePath);
+          if (!fs.existsSync(filePath)) {
+            reportDiagnostic(program, {
+              code: "fileNotFound",
+              messageId: "aboutIcon",
+              target,
+            });
+          }
+          (iconPath.type as StringLiteral).value = filePath;
+        }
+      }
+    }
+  }
+  program.stateMap(PortalCoreKeys.about).set(target, options);
+}
+
+export function getAbout(program: Program, target: Type) {
```

Comment:
Most of these functions that return things from the stateMap could use type annotations since our typing around stateMap is very weak in the compiler.

Reply:
question on this